### ISSUE TITLE
[Premium] Linux host with VNet integration to storage account

### ISSUE THREAD

**Issue by vforteli** [Azure/Azure-Functions#1417]
Linux functions do not appear to use the VNet service endpoint for accessing storage account.

I have created a http triggered Linux function which in code accesses a storage account. The linux function is connected to a VNet which has been allowed access to the storage account. The VNet also has a service endpoint for Microsoft.Storage

I would expect this function to be able to access the storage account but it ends up with error 403 forbidden. Allowing all networks on the storage account resolves this issue.

I know this is a preview feature but the VNet integration works correctly if I deploy a similar environment with a Windows based function. Also, using regular AppServices, both Windows and Linux hosts can successfully access the storage account through a VNet.

Should Linux functions be able to use VNet integrations to storage accounts at the moment?

**Comment by erik-forsmyr-stratsys** [comment]
Im not completely sure that this is the same thing but the box for vnet integration in below link is not checked for "App service linux". And the function is almost the same as a webapp in this aspect? We got the recommendation to use the ASE (Application service environment) for our web app but the pricing is very different from vnet integration.

https://azure.microsoft.com/en-us/pricing/details/app-service/plans/

**Comment by vforteli** [comment]
That page seems a bit outdated and probably only reflects GA services. For example python in functions on linux has been GA since August

**Comment by jeffhollan** [comment]
Thanks for flagging @vforteli - can you confirm how you are trying to access the storage account?  There is the Azure Functions storage account that is used to do things like publish the user code (represented in the `WEBSITE_FILECONNECTIONSTRING` and `WEBSITE_FILECONTENTSHARE` App Settings (most often the same account used in `AzureWebJobsStorage` that is created by default).  We do not have a way to support accessing these behind a VNet today (but is something we are asking).  However if your Function is just using the Storage SDK or the Storage output bindings (or even triggers) to a storage account that *isn’t* the same one as the one used for code artifacts I would expect it to work.  Can you confirm that is the case?

**Comment by vforteli** [comment]
This is happening from user code inside the function. I tried the latest v12 blob storage sdk client for NetCore and Python. 
NetCore and python functions on linux failed with 403, But the same NetCore function deployed to an otherwise similar windows function succeeded.

I didnt have any restrictions in place on the storage account used by the function host

**Comment by jeffhollan** [comment]
Ok this helps.  I’ll see if I can repro later this week.  I have done this before last month on a linux plan and it worked - so I believe it should work.  The other note is there is a bug that should be fixed in the coming weeks that sometimes it can take like 12+ hours for the routes to a service endpoint subnet to propagate through the plan.  So also worth noting you may need to wait a while after adding the subnet to the service endpoint before traffic will route to it (as mentioned it’s getting patched in current release that is rolling out before EOY).

**Comment by anindya5** [comment]
I am having same issue, I put my azure python function on a vnet in premium plan and added the vnet into my sql db, but while trying to access the db from my function using pyodbc getting error that the ip is not lsited, i.e. I have to whitelist IP addresses. That's not what I thought needs to be done once you add the vnet to the DB. Any solution?

**Comment by vforteli** [comment]
@jeffhollan yes I saw the 12 hours waiting time mentioned somewhere, but waiting doesnt seem to help here. Now its been configured like this for much longer and it didnt make any difference.

@anindya5 I was going to try Azure SQL next, simply because it tells you which IP address you appear to be coming from in the error message. 
Did you happen to check if the IP you had to whitelist was contained in the possible outbound IP addresses of the function app? For app services this seems to be the case without VNet configured, but for linux functions its apparently some completely different IP...

**Comment by vforteli** [comment]
Ok, it does actually seem to work, but the wait time was not "a few hours" or 12 hours, but more along the lines of over 24 hours 😮

**Comment by vforteli** [comment]
Tried some more stuff. After it started working (some 24 hours later) I removed the VNet integration from the function app and verified that it didnt work anymore. Now when re-adding the VNet integration, that routing was working immediately.
Next I will probably try to see if it makes a difference if the function app is immediately created with a VNet integration, instead of adding it after it has been created.

**Comment by anindya5** [comment]
@jeffhollan @vforteli - it still does not work for me. Any tips? 

@vforteli - did you switch  off , "Allow azure services and  resources to access this server" when you tried Azure SQL. If thats on, anyone can connect to the DB from any azure service. Ideally vnet integration should take care  of this, that  even with that turned off, it should  allow for the function to talk to azure sql.

**Comment by anindya5** [comment]
White listing the outbound IP ranges do work, but  always there is chance of  that changing in future, so not an acceptable solution

**Comment by vforteli** [comment]
@anindya5 I didnt have "allow azure service checked" when testing, only one rule allowing the VNet I had created for the function. Anyway, it started working at the same time as the storage account, but as mentioned, it took over 24 hours

**Comment by anindya5** [comment]
Yeah, so Azure SQL from Azure function in same Vnet did not work  for me yet by just adding the same vnet to DB. However, whitelisting the outbound IP still works for me

**Comment by vforteli** [comment]
I ran into this now with another subscription... There waiting over the weekend did not work. The subscriptions are not identical, but the setup with function, app service plan and storage account are.

**Comment by anindya5** [comment]
@alexkarcher-msft are you guys aware of this issue?

**Comment by alexkarcher-msft** [comment]
Hi @vforteli we're working on repoing and triaging this issue on our side. We have resolved the VNET integration delay issue we were tracking a few months ago, so it's taking longer to identify this new issue

@anindya5 how are you hosting your SQL DB and connecting it to your virtual network? There are a lot of SKUs of SQL and I'd like to make sure you're running into the same issue.

**Comment by anindya5** [comment]
@alexkarcher-msft created a sql server, added the db and then in the add virtual network, added the vnet. Also, in function, went to networking and added the vnet. That's it

**Comment by JarroVGIT** [comment]
Hi @alexkarcher-msft , we are running in this issue as well. You state "We have resolved the VNET integration delay issue we were tracking a few months ago" but in my case it also takes a long time for this integration to take effect. My setup: Python Function with code to access a storage account other than the hosting plan's storage account. Function works if I open up the storage account but fails if I open it up to selected networks only (and of course selecting the network/subnet that is the integration subnet used for the Function app). I am using a EP1 hosting plan.

Also, because we are bound to using Python, and subsequently are therfore bou d to using Linux, is there any ETA on GA on this functionality on Linux plans? Thanks!
